### PR TITLE
KEYCLOAK-16908 Refactor UserSessionPersisterProvider

### PR DESCRIPTION
**Pull request by martin-kanis**

### CODE REVIEW

**Comment by martin-kanis** [comment]
https://keycloak-jenkins.com/view/pipelines/job/universal-test-pipeline-server/2535
https://keycloak-jenkins.com/view/pipelines/job/universal-test-pipeline-adapters/2047

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionInitializerTest.java [hunk 1]
```diff
@@ -210,13 +218,29 @@ public void testUserSessionInitializerWithDeletingClient(KeycloakSession session
             KeycloakSession currentSession = inheritClientConnection(session, createSessionPersister3);
             RealmModel realm = currentSession.realms().getRealmByName(realmName);
 
-            // Delete cache (persisted sessions are still kept)
-            currentSession.sessions().onRealmRemoved(realm);
+            // Delete user/client[offline] cache (persisted sessions are still kept)
+            InfinispanConnectionProvider provider = currentSession.getProvider(InfinispanConnectionProvider.class);
```
Comment (mposolda) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionInitializerTest.java:222]: I am not sure if it is better to keep all the logic in the test or rather make sure that test is able to call InfinispanUserSessionProvider.removeLocalUserSessions, which is private method? WDYT @hmlnarik . My vote is to rather call the InfinispanUserSessionProvider to remove some logic from the test. Maybe we can just make that method on InfinispanUserSessionProvider public with the comment to have it public for testing purposes only? Another option is reflection, but that is not so ideal as well...

**Review by mposolda — CHANGES_REQUESTED** [review]
@martin-kanis Thanks for the PR. Changes looks good to me. I am adding one comment inline regarding the test, which I am not sure what is better, so waiting also for feedback from @hmlnarik .

Another thing is, that we can maybe remove the UserSessionPersisterProviderFactory.postInit and make sure that persister.onUserRemoved is also called from InfinispanUserSessionProviderFactory.postInit .

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionInitializerTest.java (hunk 1 above)
Comment (hmlnarik) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionInitializerTest.java:222]: Yup, I agree. That sounds as a better option, thanks for spotting this!

Comment (martin-kanis) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionInitializerTest.java]: Thank you Marek for the review. I implemented both your suggestions.

**Comment by mposolda** [comment]
@martin-kanis @hmlnarik  Thanks!

I am sorry that I forgot about one minor point - the javadoc on the UserSessionProvider.removeExpired(RealmModel) methods. As we discuss on the meeting few days ago, the current javadoc is outdated now as it mentions that calls are not propagated to persister, which is not true anymore with your changes. Could you please update the javadoc?

**Comment by martin-kanis** [comment]
> @martin-kanis @hmlnarik Thanks!
> 
> I am sorry that I forgot about one minor point - the javadoc on the UserSessionProvider.removeExpired(RealmModel) methods. As we discuss on the meeting few days ago, the current javadoc is outdated now as it mentions that calls are not propagated to persister, which is not true anymore with your changes. Could you please update the javadoc?

Good catch! Updated.

File: model/jpa/src/main/java/org/keycloak/models/jpa/session/JpaUserSessionPersisterProviderFactory.java [hunk 2]
```diff
@@ -43,6 +44,11 @@ public void init(Config.Scope config) {
 
     }
 
+    @Override
```
Comment (hmlnarik) [inline model/jpa/src/main/java/org/keycloak/models/jpa/session/JpaUserSessionPersisterProviderFactory.java:47]: I suggest also removing the `default` method from the `UserSessionPersisterProviderFactory`.

**Review by hmlnarik — CHANGES_REQUESTED** [review]
Overall looks pretty good, thanks @martin-kanis! Please see inline for two minor items

File: model/jpa/src/main/java/org/keycloak/models/jpa/session/JpaUserSessionPersisterProviderFactory.java (hunk 2 above)
Comment (martin-kanis) [reply model/jpa/src/main/java/org/keycloak/models/jpa/session/JpaUserSessionPersisterProviderFactory.java:47]: The `UserSessionPersisterProviderFactory` is empty. Which `default` do you mean?

File: server-spi/src/main/java/org/keycloak/models/UserSessionProvider.java [hunk 3]
```diff
@@ -138,8 +138,14 @@ UserSessionModel createUserSession(String id, RealmModel realm, UserModel user,
     void removeUserSession(RealmModel realm, UserSessionModel session);
     void removeUserSessions(RealmModel realm, UserModel user);
 
-    /** Implementation doesn't need to propagate removal of expired userSessions to userSessionPersister. Cleanup on persister will be called separately **/
+    /**
+     * Removal of the expired {@link UserSessionModel user sessions} is also propagated to {@link UserSessionProvider}
```
[the remark this answers is not in the document]
Comment (martin-kanis) [reply server-spi/src/main/java/org/keycloak/models/UserSessionProvider.java]: Updated

File: model/jpa/src/main/java/org/keycloak/models/jpa/session/JpaUserSessionPersisterProviderFactory.java (hunk 2 above)
Comment (hmlnarik) [reply model/jpa/src/main/java/org/keycloak/models/jpa/session/JpaUserSessionPersisterProviderFactory.java:47]: Ah, my bad, it's already done. Please ignore this comment :-)

**Review by hmlnarik — APPROVED** [review]
Looks good to me, thanks @martin-kanis!

**Review by mposolda — APPROVED** [review]
@martin-kanis @hmlnarik Thanks!

**Pull request merged**